src/shared/sleep-config.c

```diff
@@ -207,67 +296,194 @@ static bool battery_discharge_rate_is_valid(int battery_discharge_rate) {
+                        continue;
+
+                if (battery_current_capacity >= battery_last_capacity) {
+                        log_debug("Battery was not discharged during suspension");
+                        continue;
+                }
+                battery_discharge_rate = (battery_last_capacity - battery_current_capacity) * USEC_PER_HOUR / (after_timestamp - before_timestamp);
+                r = get_system_battery_identifier_hash(dev, &system_hash_id);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to generate system battery identifier hash: %m");
+
+                r = put_battery_discharge_rate(battery_discharge_rate, system_hash_id, truncate);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to update battery discharge rate, ignoring: %m");
+                else
+                        truncate = false;
+        }
+
+        return 0;
+}
+
+/* calculate the suspend interval for each battery and then return the sum of it */
+
+int get_total_suspend_interval(Hashmap *last_capacity, usec_t *ret) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        usec_t suspend_interval = 0, total_suspend_interval = 0;
+        int battery_last_capacity = 0, r;
+        sd_device *dev;
+
+        assert(last_capacity);
+        assert(ret);
+
+        r = battery_enumerator_new(&e);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to initialize battery enumerator: %m");
+
+        FOREACH_DEVICE(e, dev) {
+                const char *battery_name;
+                int previous_discharge_rate = 0;
+
+                r = sd_device_get_property_value(dev, "POWER_SUPPLY_NAME", &battery_name);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery name, ignoring: %m");
+                        continue;
+                }
+
+                battery_last_capacity = PTR_TO_INT(hashmap_get(last_capacity, battery_name));
+                if (battery_last_capacity <= 0)
+                        continue;
+
+                r = get_battery_discharge_rate(dev, &previous_discharge_rate);
+                if (r < 0) {
+                        log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_WARNING, r, "Failed to get discharge rate, ignoring: %m");
+                        continue;
+                }
+
+                if (previous_discharge_rate == 0)
+                        continue;
+
+                if (battery_last_capacity * 2 <= previous_discharge_rate) {
+                        log_debug("Current battery percentage capacity too low to suspend");
+                        continue;
+                }
+                suspend_interval = usec_sub_unsigned(battery_last_capacity * USEC_PER_HOUR / previous_discharge_rate, 30 * USEC_PER_MINUTE);
+
+                total_suspend_interval = total_suspend_interval + suspend_interval;
+        }
+
```

Comment:
if total_suspend_interval is 0 at this point, it should return -ENOENT instead of updating ret and returning 0